Framework/PythonInterface/plugins/algorithms/Abins.py

```diff
@@ -206,22 +206,44 @@ def PyExec(self):
         all_atms_smbls.sort()
 
         if len(self._atoms) == 0:  # case: all atoms
-            atoms_symbol = all_atms_smbls
+            atom_symbols = all_atms_smbls
+            atom_numbers = []
         else:  # case selected atoms
             if len(self._atoms) != len(set(self._atoms)):  # only different types
                 raise ValueError("Not all user defined atoms are unique.")
 
-            for atom_symbol in self._atoms:
+            # Specific atoms are identified with prefix and integer index, e.g 'atom_5'. Other items are element symbols
+            # A regular expression match is used to make the underscore separator optional and check the index format
+            prefix = AbinsModules.AbinsConstants.ATOM_PREFIX
+            atom_symbols = [item for item in self._atoms if item[:len(prefix)] != prefix]
+
+            numbered_atom_test = re.compile('^' + prefix + '_?(\d+)$')
+            atom_numbers = [numbered_atom_test.findall(item) for item in self._atoms]  # Matches will be lists of str
+            atom_numbers = [int(match[0]) for match in atom_numbers if match]  # Remove empty matches, cast rest to int
+
+            for atom_symbol in atom_symbols:
                 if atom_symbol not in all_atms_smbls:
-                    raise ValueError("User defined atom not present in the system.")
-            atoms_symbol = self._atoms
+                    raise ValueError("User defined atom selection (by element) '%s': not present in the system." %
+                                     atom_symbol)
+            for atom_number in atom_numbers:
+                try:
+                    atom_number = int(atom_number)
```

Comment:
Is this needed? The `atom_numbers` already seem to have been converted to ints on line 222?

Reply:
It's really there to enable the error message but you're right that it's redundant/too late. The regex will only capture natural numbers into `atom_numbers` anyway; I think the answer here is to add another check that all the entries in "atoms" were accounted for.